README.md
# Welcome to deseo

`deseo` is a Haskell library for working with Asterix data.

## Development

Development is done inside a nix shell that provides all the dependencies. To
enter run `nix-shell` at the project root - this requires
[nix](https://nixos.org/nix/) on your machine. The rest of the section assumes
you will be running commands inside the nix shell.

### Building

Use regular cabal commands (cabal is already provided by nix)

    cabal configure
    cabal build -j

If you require profiling you can obtain an environment with libraries build
with profiling by running (directly on your machine, not in the dev shell)

    nix-shell -A withProfiling

and then in the development shell

    cabal configure --enable-library-profiling
    cabal build -j

### Testing

Use regular cabal test command (cabal is already provided by nix)

    cabal test --test-option=--color
    cabal test --test-option=--color --test-show-details=always
    cabal test --test-option=--color --test-show-details=always --test-option=--maximum-generated-tests=10000 --test-option=--maximum-unsuitable-generated-tests=10000000

### Generate documentation

Use regular cabal haddock command (cabal is already provided by nix)

    cabal haddock

### Changing dependencies

Nix packaging of this library is done by having an auto-generated `default.nix`
file and a manually maintained `shell.nix`. After changing dependencies in
`deseo.cabal` you will need to run the following command to update
`default.nix`:

    cabal2nix . > default.nix

If you don't have [cabal2nix](https://github.com/NixOS/cabal2nix) on your
machine use `nix-shell -p cabal2nix` to get a temporary shell with it
available.

If you need to provide system dependencies then edit `mkEnv` function in
`shell.nix`. The `buildInputs` part defines system packages provided in the
development shell.

## Authors

This library is written and maintained by Zoran Bošnjak,
<zoran.bosnjak@sloveniacontrol.si>.



changelog.md
# 1.0

* Initial version

# 1.1

* Minor modifications for ghc801

# 1.2

* NFData instances fix

# 1.3.0.0

* build-depends update

# 1.4.1

* minor build fix



tests/Main.hs
module Main where

import Test.Framework (defaultMain)
import TestAsterix (testAsterix)
import TestBitString (testBitString)
import TestExpression (testExpression)


main :: IO ()
main = defaultMain tests where
    tests = [ testAsterix
            , testBitString
            , testExpression
            ]


tests/printout.hs
import System.Environment (getProgName, getArgs)

import Data.List
import Data.Monoid

import Data.Maybe

import Control.Monad
import Control.Exception (evaluate)
import Control.DeepSeq (force)
import Control.Monad.State

import qualified Data.ByteString as S

import qualified Data.Asterix as A
import qualified Data.Bits as B

main = do
    prog <- getProgName
    args <- getArgs

    -- read all files, force evaluation
    s <- sequence $ map readFile args
    uaps <- evaluate . force . map (\(cat, (ed, dsc)) -> (cat, dsc)) 
        . A.getCategoryDescriptions [] $ s

    let p018 :: A.Desc
        p018 = fromJust $ A.getUapByName 18 uaps "uap"
        p018_036 = A.getDesc' p018 ["036"]
        p018_036_SAC = A.getDesc' p018 ["036", "SAC"]
        p018_006 = A.getDesc' p018 ["006"]
        p018_008 = A.getDesc' p018 ["008"]
        p048 = fromJust $ A.getUapByName 48 uaps "uap"
        p062 = fromJust $ A.getUapByName 62 uaps "uap"
        p062_105_LAT = A.getDesc' p062 ["105","LAT"]

    let 
        record0 = A.create p048 $ return ()
        
        record1a = A.Item p048 $ B.bits 24 0x800102
        record1b = A.Item p048 $ B.bits 32 0x40010203

        record2 = A.create p048 $ do
            A.setItem "010" $ A.fromValue 0x0102

        record3a = A.create p018 $ do
            A.setItem "036" $ A.fromValue 0x0102

        record3b = A.create p018 $ do
            A.setItem "037" $ A.fromValue 0x0102

        record3c = A.create p018 $ do
            A.setItem "036" $ A.fromValue 0x0102
            A.setItem "037" $ A.fromValue 0x0304

        record3d = A.create p018 $ do
            A.setItem "036" $ A.fromValue 0x0102
            A.setItem "037" $ A.fromValues [("SAC", 0x03), ("SIC", 0x04)]

        record4 = A.create p018 $ do
            A.setItem "006" $ A.fromList 
                [ [("ModeS", 0x010203)]
                , [("ModeS", 0x020304)]
                ]
        
        record5 = A.create p048 $ do 
            A.setItem "130" $ A.fromValues [("SRL", 0xaa), ("SRR", 0xbb)]

        record6 = A.Item p048 $ B.bits 24 0x20FF8E

        item7a = A.Item p018_008 $ B.bits 8 0x00
        item7b = A.Item p018_008 $ B.bits 16 0x0100

        record8 = A.create p062 $ do
            {-
                /105/LAT Fixed (32) 008238ca -> 45.7811129093 deg
                /105/LON Fixed (32) 001d6da2 -> 10.3458702564 deg
                /070 Fixed (24) 2dbacc -> 23413.59375 s
                /290 Compound (32) 70ff1010
                    /290/PSR Fixed (8) ff
                    /290/SSR Fixed (8) 10
                    /290/MDS Fixed (8) 10
            -}
            A.setItem "105" $ A.fromValues
                [("LAT", 0X008238ca), ("LON", 0x001d6da2)]
            A.setItem "070" $ A.fromValue 0x2dbacc
            -- A.setItem "290" $ A.fromValue 0x70ff1010
            -- A.setItem "290" $ A.fromValues 
            --  [("PSR", 0xff), ("SSR", 0x10), ("MDS", 0x10)]

    print $ A.sizeOf p018_036 $ B.bits 100 0
    print $ A.sizeOf p018_036_SAC $ B.bits 100 0
    print $ A.sizeOf p018_036_SAC $ B.bits 1 0
    print p018_006
    print $ A.sizeOf p018_006 $ B.bits 100 0
    print $ A.sizeOf p018_006 $ B.bits 32 0x01000000
    print $ A.sizeOf p018_006 $ B.bits 64 0x0200000000000000
    print $ A.sizeOf p018 $ B.bits 24 0x800102

    print record0
    print $ A.childsComp record0
    print record1a
    print $ A.childsComp record1a
    print record1b
    print $ A.unChildsComp p048 . A.childsComp $ record1b
    print $ A.childsComp record1b
    print record2
    print record3a
    print record3b
    print record3c
    print record3d
    print record4
    print record5
    print $ A.encodeDb $ A.datablock 48 [record5]
    print $ A.childs . fromJust $ return record2 >>= A.child "010"
    print $ return record2 >>= A.child "010" >>= A.child "SAC"
    print $ A.childR ["010","SAC"] record2
    print $ A.toValue . fromJust $ A.childR ["010","SAC"] record2
    print $ A.childs . fromJust $ return record6 >>= A.child "020"
    print $ 
        map A.toValue $ A.childs . fromJust $ return record6 >>= A.child "020"
    print $ A.childs item7a
    print $ A.childs item7b
    print $ p062_105_LAT
    print "---"
    print record8
    print $ return record8 >>= A.child "105" >>= A.child "LAT"
    print $ return record8 >>= A.child "105" >>= A.child "LAT" >>= A.getFloating
    print $ return record8 >>= A.child "105" >>= A.child "LON" >>= A.getFloating
    print $ return record8 >>= A.child "070" >>= A.getFloating
    print $ return record8 >>= A.child "070" >>= A.getRaw
    --print $ return record8 >>= A.child "290"
    --print $ return record8 >>= A.child "290" >>= A.getRaw

    return ()



Bits.hs

{-

Bit string manipulation
(low performance version, based on list)

Author: Zoran Bosnjak (Sloveniacontrol)

-}

module Bits
(   Bits
    , bits
    , length
    , take
    , drop
    , index
    , zeros
    , pack
    , unpack
    , null
    , checkAligned
    , toUnsigned
    , fromByteString
    , toByteString
) where

import Control.Monad
import qualified Prelude as P
import Prelude hiding (length, any, toInteger, take, drop, null, (!!))
import Data.Monoid
import qualified Data.ByteString as S
import Data.Word
import qualified Data.Bits as B
import Data.List (foldl', foldl1')
import Data.Maybe

import Debug.Trace

import Test.QuickCheck

dump = flip trace

data Bits = Bits [Bool] 
                deriving (Eq)

instance Show Bits where
    show (Bits b) = "Bits " ++ disp b where
        disp [] = ""
        disp x = foldl1' (\a b -> a++" "++b) . spl . map f $ x
        f True = '1'
        f False = '0'
        spl [] = []
        spl s = 
            let (a,b) = splitAt 8 s
            in (fill a):(spl b)
        fill a = P.take 8 (a++repeat '.')

instance Monoid Bits where
    mempty = Bits []
    Bits a `mappend` Bits b = Bits (a `mappend` b)

instance Arbitrary Bits where
    arbitrary = liftM Bits arbitrary

length (Bits a) = P.length a
take n (Bits a) = Bits $ P.take n a
drop n (Bits a) = Bits $ P.drop n a
index (Bits b) n = (P.!!) b n
pack = Bits
unpack (Bits b) = b
null (Bits b) = P.null b

-- generate bitstring
bits :: Integral a => Int -> a -> Bits
bits n val = Bits . map toBool $ f n val [] where
    f 0 _ acc = acc
    f n val acc = f (n-1) a (b:acc) where
        (a,b) = divMod val 2
    toBool 0 = False
    toBool 1 = True

zeros n = bits n 0

-- is byte aligned
checkAligned :: Bits -> Maybe Bits
checkAligned b = case (length b `mod` 8) of
    0 -> Just b
    _ -> Nothing

-- False=0, True=1
boolVal :: Num a => Bool -> a
boolVal False = 0
boolVal True = 1

-- convert bits to unsigned number
toUnsigned :: Num a => Bits -> a
toUnsigned (Bits b) = sum . zipWith (*) factors . map boolVal . reverse $ b where
    factors = map (2^) [0..]

fromByteString :: S.ByteString -> Bits
fromByteString = Bits . concatMap octet . S.unpack where
    octet w = map (B.testBit w) [7,6..0]

toByteString :: Bits -> Maybe S.ByteString
toByteString bs = do
    (Bits s) <- checkAligned bs
    return . S.pack . map toWord . break8 $ s where
        break8 :: [a] -> [[a]]
        break8 [] = [] 
        break8 s = a : break8 b where (a,b) = splitAt 8 s
        toWord :: [Bool] -> Word8
        toWord s = sum $ zipWith (*) (map (2^) [7,6..0]) (map boolVal s)

-- run the tests
main :: IO ()
main = do
    putStrLn "quick check test run..."

    quickCheck testPack
    quickCheck testPack2
    quickCheck testByteString
    quickCheck testUnsigned
    quickCheck testCombine
    quickCheck testZeros

    where

    testPack :: [Bool] -> Bool
    testPack s =    ((unpack . pack $ s) == s)
                    && (pack s) == (pack . unpack . pack $ s)

    testPack2 :: Bits -> Bool
    testPack2 b = (pack . unpack $ b) == b

    testByteString :: [Word8] -> Bool
    testByteString s = (fromJust . toByteString . fromByteString . S.pack $ s) == (S.pack s)

    testUnsigned :: NonNegative Int -> Bool
    testUnsigned (NonNegative val) = toUnsigned (bits (n+1) val) == val where
        n = ceiling . logBase 2 $ fromIntegral val

    testCombine :: Bits -> Bits -> Bool
    testCombine a b =   (pack (unpack a ++ unpack b) == c)
                        && (take (length a) c == a)
                        && (drop (length a) c == b)
                        where c = a `mappend` b
    
    testZeros :: Property
    testZeros = forAll (choose (0,100)) $ \n -> 
         (toUnsigned . zeros $ n) == 0



test1.hs
#!/usr/bin/env runhaskell

import System.Environment (getProgName, getArgs)

import Data.List
import Data.Monoid
import Data.Maybe

import Control.Monad
import Control.Exception (evaluate)
import Control.DeepSeq (force)

import qualified Data.ByteString as S

import qualified Asterix as A

main = do
    prog <- getProgName
    args <- getArgs

    -- read all files, force evaluation
    s <- sequence $ map readFile args
    let requested = [ (2, A.Edition 1 0)
                    ]
    profiles <- evaluate . force . A.getCategoryDescriptions requested $ s

    forM_ profiles $ \(cat, (edition, dsc)) -> do
        putStrLn $ show cat
        putStrLn $ "  " ++ show edition


    -- get cat002
    let (_, uaps) = fromJust $ lookup 2 profiles
        dsc = A.getUap 2 uaps mempty
        d010 = head . A.dItems $ dsc
        d000 = head . tail . A.dItems $ dsc
        d050 = head . tail . tail . tail . tail . tail . A.dItems $ dsc
        d070 = head . tail . tail . tail . tail . tail . tail . tail . A.dItems $ dsc
        dSP = head . tail . tail . tail . tail . tail . tail . tail . tail . tail . tail . tail . tail . A.dItems $ dsc
        bs = A.Bytes (S.pack [0x02, 0x03])
        bs2 = A.Bytes (S.pack [0x03, 0x05, 0x02])
        bs3 = A.Bytes (S.pack [0x02, 0x01, 0x02, 0x03, 0x04])
        bs4 = A.Bytes (S.pack [0x40, 0x01, 0x02, 0x03])

        i010 = A.decode d010 bs
        i000 = A.decode d000 bs
        i050 = A.decode d050 bs2
        i070 = A.decode d070 bs3
        iSP  = A.decode dSP bs
        root = A.decode dsc bs4

    putStrLn ""

    putStrLn . show $ d010
    putStrLn . show $ i010

    putStrLn . show $ d000
    putStrLn . show $ i000

    putStrLn . show $ d050
    putStrLn . show $ i050

    putStrLn . show $ d070
    putStrLn . show $ i070

    putStrLn . show $ dSP
    putStrLn . show $ iSP

    putStrLn . show $ dsc
    putStrLn . show $ root



Data/BitString.hs
----------------
-- |
-- Module       :  Data.BitString
-- Copyright:   (c) 2015-2016 Zoran Bošnjak
--              (c) 2015-2016 Sloveniacontrol Ltd. (www.sloveniacontrol.si)
-- License:     GPL-3
-- Maintainer:  Zoran Bošnjak <zoran.bosnjak@sloveniacontrol.si>
--
-- This module provides simple bitstring manipulation library.
-- The implementation is based on lists - [Bool].
--
-- This module is intended to be imported qualified, e.g.
--
-- > import qualified Data.BitString as B
--
-- Examples:
--
-- >    >>> B.pack [True, False, False]
-- >    Bits 100.....
-- >    >>> B.take 2 $ B.pack [True, False, False]
-- >    Bits 10......
-- >    >>> B.fromInteger 16 0x0102
-- >    Bits 00000001 00000010
-- >    >>> B.toUIntegral $ B.pack [True, False]
-- >    2
--

{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}

module Data.BitString (

    Bits

    -- * Bits creation
    , pack

    -- * Util functions
    , length
    , take
    , takeMaybe
    , drop
    , dropMaybe
    , index
    , zeros
    , unpack
    , null
    , checkAligned
    , complement
    , anySet, allSet

    -- * Convert functions
    , fromInteger
    , toSIntegral, toUIntegral
    , fromByteString
    , toByteString
) where

import Control.DeepSeq
import Control.Monad
import qualified Prelude as P
import Prelude
    hiding (length, any, fromInteger, toInteger, take, drop, null, (!!))
import qualified Data.ByteString as S
import Data.Word
import qualified Data.Bits as B
import Data.List (unfoldr)
import GHC.Generics

import Test.QuickCheck

-- | Bits data type.
newtype Bits = Bits [Bool] deriving (Generic, Eq, Semigroup, Monoid, Arbitrary)

instance NFData Bits

instance Show Bits where
    show (Bits bb) = "Bits " ++ unwords octets where
        octets = unfoldr getOctet bb
        getOctet [] = Nothing
        getOctet s =
            let (a,b) = splitAt 8 s
                octet = P.take 8 ((bitValue <$> a) ++ repeat '.')
            in Just (octet, b)
        bitValue True = '1'
        bitValue False = '0'

-- | Calculate length of bitstring.
length :: Bits -> Int
length (Bits a) = P.length a

-- | Take first 'n' bits.
take :: Int -> Bits -> Bits
take n (Bits a) = Bits $ P.take n a

-- | Take first 'n' bits (if available).
takeMaybe :: Int -> Bits -> Maybe Bits
takeMaybe n b
    | length b < n = Nothing
    | otherwise = Just $ take n b

-- | Drop first 'n' bits.
drop :: Int -> Bits -> Bits
drop n (Bits a) = Bits $ P.drop n a

-- | Drop first 'n' bits (if available)
dropMaybe :: Int -> Bits -> Maybe Bits
dropMaybe n b
    | length b < n = Nothing
    | otherwise = Just $ drop n b

-- | Return bit at given index.
index :: Bits -> Int -> Bool
index (Bits b) n = b P.!! n

-- | Convert from list of Bool values to Bits
pack :: [Bool] -> Bits
pack = Bits

-- | Convert from Bits to list of Bool values
unpack :: Bits -> [Bool]
unpack (Bits b) = b

-- | Is empty bitstring?
null :: Bits -> Bool
null (Bits b) = P.null b

-- | Return 'n' bits of all zero values.
zeros :: Int -> Bits
zeros n = Bits $ replicate n False

-- | Generate bitstring of given bit length and value,
-- for example:
--
-- >    fromInteger 16 0x1234 == Bits 00010010 00110100
--
fromInteger :: Int -> Integer -> Bits
fromInteger n val = Bits $ reverse $ unfoldr f (n,val) where
    f (0,_) = Nothing
    f (n', val') =
        let (a,b) = val' `divMod` 2
        in Just (bitValue b, (n'-1, a))
    bitValue 0 = False
    bitValue _ = True

-- | If byte aligned, return Just value, else 'Nothing'.
checkAligned :: Bits -> Maybe Bits
checkAligned b = case (length b `mod` 8) of
    0 -> Just b
    _ -> Nothing

-- False=0, True=1
boolVal :: Num a => Bool -> a
boolVal False = 0
boolVal True = 1

-- | Convert bits to signed number.
toSIntegral :: Num a => Bits -> a
toSIntegral b
    | null b = 0
    | (head . unpack $ b) == False = toUIntegral b
    | otherwise = -((toUIntegral $ complement b)+1)

-- | Convert bits to unsigned number.
toUIntegral :: Num a => Bits -> a
toUIntegral (Bits b) = sum . zipWith (*) factors . map boolVal . reverse $ b
  where
    factors = map (2^) ([0..]::[Int])

-- | Convert bytestring to Bits.
fromByteString :: S.ByteString -> Bits
fromByteString = Bits . concatMap octet . S.unpack where
    octet w = map (B.testBit w) [7,6..0]

-- | Convert Bits to bytestring.
toByteString :: Bits -> Maybe S.ByteString
toByteString bs = do
    (Bits s) <- checkAligned bs
    return . S.pack . map toWord . break8 $ s where
        break8 :: [a] -> [[a]]
        break8 [] = []
        break8 s = a : break8 b where (a,b) = splitAt 8 s
        toWord :: [Bool] -> Word8
        toWord s = sum $
            zipWith (*) (map (2^) ([7,6..0]::[Int])) (map boolVal s)

-- | Reverse all the bits.
complement :: Bits -> Bits
complement (Bits b) = Bits $ map not b

-- | Is any bit set?
anySet :: Bits -> Bool
anySet = P.any id . unpack

-- | Are all bits set?
allSet :: Bits -> Bool
allSet = P.all id . unpack


Data/Asterix/Expression.hs
----------------
-- |
-- Module:      Data.Asterix.Expression
-- Copyright:   (c) 2015-2016 Zoran Bošnjak
--              (c) 2015-2016 Sloveniacontrol Ltd. (www.sloveniacontrol.si)
-- License:     GPL-3
-- Maintainer:  Zoran Bošnjak <zoran.bosnjak@sloveniacontrol.si>
--
-- This module provides python like math expression evaluator
-- for some basic expressions.
--

{-# LANGUAGE DeriveGeneric #-}

module Data.Asterix.Expression
( EValue(EInteger, EDouble)
, eval
) where

import Control.Applicative ((<|>))
import Control.DeepSeq (NFData)
import Control.Monad (mzero, void)
import GHC.Generics (Generic)
import Text.Megaparsec (between, parseMaybe, try, Parsec)
import Text.Megaparsec.Char (spaceChar)
import Text.Megaparsec.Expr (makeExprParser, Operator(InfixL))
import Data.Void (Void)
import qualified Text.Megaparsec.Char.Lexer as L

type Parser = Parsec Void String

data EValue
    = EInteger Integer
    | EDouble Double
    deriving (Show, Generic)

instance NFData EValue

-- | Convert to Double
_toDouble :: EValue -> Double
_toDouble (EInteger x) = fromIntegral x
_toDouble (EDouble x) = x

instance Num EValue where

    -- (+)
    EInteger a + EInteger b = EInteger $ a+b
    a + b = EDouble $ (_toDouble a)+(_toDouble b)

    -- (*)
    EInteger a * EInteger b = EInteger $ a*b
    a * b = EDouble $ (_toDouble a)*(_toDouble b)

    -- abs
    abs (EInteger x) = EInteger $ abs x
    abs (EDouble x) = EDouble $ abs x

    -- signum
    signum (EInteger x) = EInteger $ signum x
    signum (EDouble x) = EDouble $ signum x

    -- fromInteger
    fromInteger = EInteger

    -- negate
    negate (EInteger x) = EInteger $ negate x
    negate (EDouble x) = EDouble $ negate x

instance Fractional EValue where
    fromRational = EDouble . fromRational
    recip (EInteger x) = EDouble . recip $ fromIntegral x
    recip (EDouble x) = EDouble $ recip x

instance Eq EValue where
    EInteger val1 == EInteger val2 = val1 == val2
    val1 == val2 = _toDouble val1 == _toDouble val2

instance Ord EValue where
    compare (EInteger val1) (EInteger val2) = compare val1 val2
    compare val1 val2 = compare (_toDouble val1) (_toDouble val2)


data Expr
  = EValue EValue
  | EPow Integer Integer
  | EBinary BinOp Expr Expr
  deriving (Show)

data BinOp
  = Add
  | Subtract
  | Multiply
  | Divide
  deriving (Show)

-- sc stands for "space consumer" as per Megaparsec convention
sc :: Parser ()
sc = L.space (void spaceChar) mzero mzero

symbol :: String -> Parser String
symbol = L.symbol sc

parened :: Parser a -> Parser a
parened = between (symbol "(") (symbol ")")

integer :: Parser Integer
integer = L.signed sc $ hex <|> L.decimal where
    hex = (try $ symbol "0x") >> L.hexadecimal

double :: Parser Double
double = L.signed sc L.float

evalue :: Parser EValue
evalue = try (EDouble <$> double) <|> (EInteger <$> integer)

pow :: Parser Expr
pow = do
    void $ symbol "pow"
    parened $ do
        a <- integer
        void $ symbol ","
        b <- integer
        return $ EPow a b

binOp :: Parser Expr
binOp = makeExprParser term
  [ [ InfixL (EBinary Multiply <$ symbol "*")
    , InfixL (EBinary Divide   <$ symbol "/") ]
  , [ InfixL (EBinary Add      <$ symbol "+")
    , InfixL (EBinary Subtract <$ symbol "-") ]
  ]

term :: Parser Expr
term = parened expr <|> pow <|> (EValue <$> evalue)

expr :: Parser Expr
expr = binOp <|> term


-- | Parse and evaluate string expression.
-- Examples:
--
-- >    > eval "1+2"
-- >    Just (EInteger 3)
-- >
-- >    > eval "pow(2,10)"
-- >    Just (EInteger 1024)
-- >
-- >    > eval "-180.0/pow(0x02, 23)"
-- >    Just (EDouble (-2.1457672119140625e-5))
--
-- >    eval "pow(10,-2)"
-- >    Just (EDouble 1.0e-2)
-- >
-- >    > eval "test"
-- >    Nothing
--
eval :: String -> Maybe EValue
eval s = compute <$> parseMaybe expr s

compute :: Expr -> EValue
compute (EValue e) = e
compute (EPow m e) = if (e>=0)
                     then EInteger $ m^e
                     else EDouble $ 1.0 / (fromInteger (m^(-e)))
compute (EBinary op a b) =
    let f = case op of Add -> (+)
                       Subtract -> (-)
                       Multiply -> (*)
                       Divide -> (/)
    in  f (compute a) (compute b)


test3.hs
#!/usr/bin/env runhaskell

import System.Environment (getProgName, getArgs)

import Data.List
import Data.Monoid

import Data.Maybe

import Control.Monad
import Control.Exception (evaluate)
import Control.DeepSeq (force)
import Control.Monad.State

import qualified Data.ByteString as S

import qualified Asterix as A
import qualified Bits as B

main = do
    prog <- getProgName
    args <- getArgs

    -- read all files, force evaluation
    s <- sequence $ map readFile args
    uaps <- evaluate . force . map (\(cat, (ed, dsc)) -> (cat, dsc)) . A.getCategoryDescriptions [] $ s

    let p018 :: A.Desc
        p018 = fromJust $ A.getUapByName 18 uaps "uap"
        p018_036 = A.getDesc' p018 ["036"]
        p018_036_SAC = A.getDesc' p018 ["036", "SAC"]
        p018_006 = A.getDesc' p018 ["006"]
        p018_008 = A.getDesc' p018 ["008"]
        p048 = fromJust $ A.getUapByName 48 uaps "uap"
        p062 = fromJust $ A.getUapByName 62 uaps "uap"
        p062_105_LAT = A.getDesc' p062 ["105","LAT"]

    let 
        record0 = A.create p048 $ return ()
        
        record1a = A.Item p048 $ B.bits 24 0x800102
        record1b = A.Item p048 $ B.bits 32 0x40010203

        record2 = A.create p048 $ do
                    A.setItem "010" $ A.fromValue 0x0102

        record3a = A.create p018 $ do
                    A.setItem "036" $ A.fromValue 0x0102

        record3b = A.create p018 $ do
                    A.setItem "037" $ A.fromValue 0x0102

        record3c = A.create p018 $ do
                    A.setItem "036" $ A.fromValue 0x0102
                    A.setItem "037" $ A.fromValue 0x0304

        record3d = A.create p018 $ do
                    A.setItem "036" $ A.fromValue 0x0102
                    A.setItem "037" $ A.fromValues [("SAC", 0x03), ("SIC", 0x04)]

        record4 = A.create p018 $ do
                    A.setItem "006" $ A.fromList [
                                            [("ModeS", 0x010203)]
                                            , [("ModeS", 0x020304)]
                                       ]
        
        record5 = A.create p048 $ do 
                    A.setItem "130" $ A.fromValues [("SRL", 0xaa), ("SRR", 0xbb)]

        record6 = A.Item p048 $ B.bits 24 0x20FF8E

        item7a = A.Item p018_008 $ B.bits 8 0x00
        item7b = A.Item p018_008 $ B.bits 16 0x0100

        record8 = A.create p062 $ do
                    {-
                        /105/LAT Fixed (32) 008238ca -> 45.7811129093 deg
                        /105/LON Fixed (32) 001d6da2 -> 10.3458702564 deg
                        /070 Fixed (24) 2dbacc -> 23413.59375 s
                        /290 Compound (32) 70ff1010
                            /290/PSR Fixed (8) ff
                            /290/SSR Fixed (8) 10
                            /290/MDS Fixed (8) 10
                    -}
                    A.setItem "105" $ A.fromValues [("LAT", 0X008238ca), ("LON", 0x001d6da2)]
                    A.setItem "070" $ A.fromValue 0x2dbacc
                    -- A.setItem "290" $ A.fromValue 0x70ff1010
                    -- A.setItem "290" $ A.fromValues [("PSR", 0xff), ("SSR", 0x10), ("MDS", 0x10)]

    print $ A.sizeOf p018_036 $ B.bits 100 0
    print $ A.sizeOf p018_036_SAC $ B.bits 100 0
    print $ A.sizeOf p018_036_SAC $ B.bits 1 0
    print p018_006
    print $ A.sizeOf p018_006 $ B.bits 100 0
    print $ A.sizeOf p018_006 $ B.bits 32 0x01000000
    print $ A.sizeOf p018_006 $ B.bits 64 0x0200000000000000
    print $ A.sizeOf p018 $ B.bits 24 0x800102

    print record0
    print $ A.childsComp record0
    print record1a
    print $ A.childsComp record1a
    print record1b
    print $ A.unChildsComp p048 . A.childsComp $ record1b
    print $ A.childsComp record1b
    print record2
    print record3a
    print record3b
    print record3c
    print record3d
    print record4
    print record5
    print $ A.encodeDb $ A.datablock 48 [record5]
    print $ A.childs . fromJust $ return record2 >>= A.child "010"
    print $ return record2 >>= A.child "010" >>= A.child "SAC"
    print $ A.childR ["010","SAC"] record2
    print $ A.toValue . fromJust $ A.childR ["010","SAC"] record2
    print $ A.childs . fromJust $ return record6 >>= A.child "020"
    print $ map A.toValue $ A.childs . fromJust $ return record6 >>= A.child "020"
    print $ A.childs item7a
    print $ A.childs item7b
    print $ p062_105_LAT
    print "---"
    print record8
    print $ return record8 >>= A.child "105" >>= A.child "LAT"
    print $ return record8 >>= A.child "105" >>= A.child "LAT" >>= A.getFloating
    print $ return record8 >>= A.child "105" >>= A.child "LON" >>= A.getFloating
    print $ return record8 >>= A.child "070" >>= A.getFloating
    print $ return record8 >>= A.child "070" >>= A.getRaw
    --print $ return record8 >>= A.child "290"
    --print $ return record8 >>= A.child "290" >>= A.getRaw

    return ()

